README.md
[![crates.io](https://img.shields.io/crates/v/slack-blocks.svg)](https://crates.io/crates/slack-blocks)
[![docs.rs](https://docs.rs/slack-blocks/badge.svg)](https://docs.rs/slack-blocks/latest)
![Maintenance](https://img.shields.io/badge/maintenance-activly--developed-brightgreen.svg)

# slack-blocks

This crate brings Slack's terrific [Block Kit 🔗] to
the Rust ecosystem.

Inside, you'll find models for all of Slack's Layout Blocks,
Block Elements, and Composition Objects. Each structure has Slack's API
documentation copied in-place so you don't have to leave your editor to
remember the details of the block kit API.

Every model has builders that leverage Rust's type system
to help you provide every required field, so you can be confident in your app.

### Troubleshooting common compiler errors
`Method build not found for ...Builder` - Dig into the error message,
you'll find something like `RequiredMethodNotCalled<method::foo>`,
meaning you need to call `.foo()` before you can call `.build()`!

## Example
Using an example from Slack's Documentation:
```json
{
  "type": "section",
  "text": {
    "text": "*Sally* has requested you set the deadline for the Nano launch project",
    "type": "mrkdwn"
  },
  "accessory": {
    "type": "datepicker",
    "action_id": "datepicker123",
    "initial_date": "1990-04-28",
    "placeholder": {
      "type": "plain_text",
      "text": "Select a date"
    }
  }
}
```

You can use raw Builders like so:
```rust
use slack_blocks::{text::ToSlackMarkdown, blocks::Section, elems::DatePicker};

let section = Section::builder()
                      .text("*Sally* has requested you set the deadline for the Nano launch project".markdown())
                      .accessory(DatePicker::builder()
                                            .action_id("datepicker123")
                                            .initial_date((28, 4, 1990))
                                            .placeholder("Select a date")
                                            .build()
                      )
                      .build();
```

Or enable the `unstable` feature and use xml macros:
```rust
use slack_blocks::blox::*;

let pick_date = blox! {
  <date_picker action_id="datepicker123"
               placeholder="Select a date"
               initial_date=(28, 4, 1990) />
};

let section = blox! {
  <section_block accessory=pick_date>
    <text kind=plain>"*Sally* has requested you set the deadline for the Nano launch project"</text>
  </section_block>
};
```

Then you can send the block to Slack's API, for example:

```rust
let blocks: Vec<Block> = vec![section.into()]; // using section from examples above

reqwest::Client::new()
                .post("https://slack.com/api/chat.postMessage")
                .header("Content-Type", "application/json")
                .bearer_auth("<api token here>")
                .body(serde_json::json!({
                  "channel": "<a channel id>",
                  "blocks": blocks
                }).to_string())
                .send();
```

There is also a crate example (`./examples/reqwest.rs`) that can be run like so:
```sh
> export SLACK_API_TOKEN=foo
> export SLACK_API_TOKEN=bar
> cargo run --all-features --example reqwest
```

[Block Kit 🔗]: https://api.slack.com/block-kit
[`cargo-make`]: https://github.com/sagiegurari/cargo-make/
[issues]: https://github.com/cakekindel/slack-blocks-rs/issues/
[Conventional Commits]: https://www.conventionalcommits.org/en/v1.0.0/

## License

Licensed under either of

* Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or https://www.apache.org/licenses/LICENSE-2.0)
* MIT license ([LICENSE-MIT](LICENSE-MIT) or https://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally
submitted for inclusion in the work by you, as defined in the Apache-2.0
license, shall be dual licensed as above, without any additional terms or
conditions.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.25.0 (2021-06-05)


### ⚠ BREAKING CHANGES

* crate should support header block and thoroughly document blox module (#165)

### Features

* crate should support header block and thoroughly document blox module ([#165](https://github.com/cakekindel/slack-blocks-rs/issues/165)) ([f57eb4d](https://github.com/cakekindel/slack-blocks-rs/commit/f57eb4d558bcc29c335ffea49c63d9c01ef56750))

### 0.24.4 (2021-06-04)

### 0.24.3 (2021-06-04)


### Bug Fixes

* html-root-updater should export readVersion ([#163](https://github.com/cakekindel/slack-blocks-rs/issues/163)) ([b482a61](https://github.com/cakekindel/slack-blocks-rs/commit/b482a61ec16d8de29616bfe7d0d78043d7044420))

### 0.24.2 (2021-06-04)


### Bug Fixes

* crate should set html_root_url and keep up to date ([#162](https://github.com/cakekindel/slack-blocks-rs/issues/162)) ([d40f2d2](https://github.com/cakekindel/slack-blocks-rs/commit/d40f2d2c08b0b280e603901fd5e7bf3e22a32cc8))

### 0.24.1 (2021-06-03)


### Bug Fixes

* validation feature should compile ([#161](https://github.com/cakekindel/slack-blocks-rs/issues/161)) ([2eb9580](https://github.com/cakekindel/slack-blocks-rs/commit/2eb95805b21fde32155db395e1582904fd7da336))

## 0.24.0 (2021-06-03)


### ⚠ BREAKING CHANGES

* move runtime validation to crate feature (#160)

### Features

* move runtime validation to crate feature ([#160](https://github.com/cakekindel/slack-blocks-rs/issues/160)) ([651b1b7](https://github.com/cakekindel/slack-blocks-rs/commit/651b1b7d04b6964e4cc1d7117974f0a78adee53d))

### 0.23.1 (2021-06-02)

## 0.23.0 (2021-06-02)


### ⚠ BREAKING CHANGES

* rename mox->blox (#159)

### Features

* rename mox->blox ([#159](https://github.com/cakekindel/slack-blocks-rs/issues/159)) ([77e98ee](https://github.com/cakekindel/slack-blocks-rs/commit/77e98ee2f969e1973c9f974ba8b9833e95d79213))

## 0.22.0 (2021-06-02)


### ⚠ BREAKING CHANGES

* remove deprecated functionality (#158)

### Features

* remove deprecated functionality ([#158](https://github.com/cakekindel/slack-blocks-rs/issues/158)) ([3078178](https://github.com/cakekindel/slack-blocks-rs/commit/3078178e63d362c14121f6c7c716db946ab3b104))

## 0.21.0 (2021-06-01)


### ⚠ BREAKING CHANGES

* all structures should serialize correctly (#157)

### Features

* all structures should serialize correctly ([#157](https://github.com/cakekindel/slack-blocks-rs/issues/157)) ([8b44945](https://github.com/cakekindel/slack-blocks-rs/commit/8b44945df32c3e1cf587f28310a776bb61199a41))

### 0.20.2 (2021-05-29)

### 0.20.1 (2021-05-28)


### Bug Fixes

* cargo had >5 keywords ([24defec](https://github.com/cakekindel/slack-blocks-rs/commit/24defec482b9811bf75743c0721215b495c6235e))

## 0.20.0 (2021-05-28)


### ⚠ BREAKING CHANGES

* xml macro support (minor breaking)

### Features

* xml macro support (minor breaking) ([51b9239](https://github.com/cakekindel/slack-blocks-rs/commit/51b9239cf1fdedab9c706364a657a027875c110d))

### 0.19.9 (2021-05-28)


### Features

* support for xml macros! ([#155](https://github.com/cakekindel/slack-blocks-rs/issues/155)) ([4717a96](https://github.com/cakekindel/slack-blocks-rs/commit/4717a96180f949ac22c5f199f6b35e57934dfdcb))

### 0.19.8 (2021-05-26)


### Features

* confirm builder ([#154](https://github.com/cakekindel/slack-blocks-rs/issues/154)) ([6d9e024](https://github.com/cakekindel/slack-blocks-rs/commit/6d9e0243147cf5a54a277b9be7c747087eb89fc2))

### 0.19.7 (2021-05-26)


### Features

* text builder ([#153](https://github.com/cakekindel/slack-blocks-rs/issues/153)) ([49e845a](https://github.com/cakekindel/slack-blocks-rs/commit/49e845a0df814c8b46cef143f619b247a39ffc52))

### 0.19.6 (2021-05-25)


### Features

* input builder ([#152](https://github.com/cakekindel/slack-blocks-rs/issues/152)) ([2c0a305](https://github.com/cakekindel/slack-blocks-rs/commit/2c0a30580ff14b5b2bb10d66c57069e8409104ba))

### 0.19.5 (2021-05-25)


### Features

* section block builder ([#151](https://github.com/cakekindel/slack-blocks-rs/issues/151)) ([ae95171](https://github.com/cakekindel/slack-blocks-rs/commit/ae95171078deb909540a61a3316ef12e818ecccc))

### 0.19.4 (2021-05-25)


### Features

* image builder ([#150](https://github.com/cakekindel/slack-blocks-rs/issues/150)) ([5d4d8b4](https://github.com/cakekindel/slack-blocks-rs/commit/5d4d8b413e716c1a9078d770dae2d40e8ab8a08e))

### 0.19.3 (2021-05-25)


### Features

* file builder ([#149](https://github.com/cakekindel/slack-blocks-rs/issues/149)) ([81fd5d7](https://github.com/cakekindel/slack-blocks-rs/commit/81fd5d70778787a86c87737aac7f2bc27e110438))

### 0.19.2 (2021-05-25)


### Features

* context builder ([#147](https://github.com/cakekindel/slack-blocks-rs/issues/147)) ([da4d125](https://github.com/cakekindel/slack-blocks-rs/commit/da4d1253442f141af9332f17f6ef6583a84f7bac))

### 0.19.1 (2021-05-25)


### Features

* actions builder ([#146](https://github.com/cakekindel/slack-blocks-rs/issues/146)) ([5bc0de0](https://github.com/cakekindel/slack-blocks-rs/commit/5bc0de050c4b9ea35b4183869ef33edb54c9e8c8))

## 0.19.0 (2021-05-24)


### ⚠ BREAKING CHANGES

* standardize supported block elements (#145)

### Features

* standardize supported block elements ([#145](https://github.com/cakekindel/slack-blocks-rs/issues/145)) ([bdbf7c7](https://github.com/cakekindel/slack-blocks-rs/commit/bdbf7c7ac198223d1682db42c729b012241a8a00))

## 0.18.0 (2021-05-24)


### ⚠ BREAKING CHANGES

* button should use cow and have builder (#144)

### Bug Fixes

* button should use cow and have builder ([#144](https://github.com/cakekindel/slack-blocks-rs/issues/144)) ([4a73b39](https://github.com/cakekindel/slack-blocks-rs/commit/4a73b39d4a1ba4984129fe8e89eb9dbda4ed6124))

### 0.17.2 (2021-05-24)

### 0.17.1 (2021-05-24)


### Bug Fixes


src/lib.rs
//! This crate brings Slack's terrific [Block Kit 🔗] to
//! the Rust ecosystem.
//!
//! Inside, you'll find models for all of Slack's Layout Blocks,
//! Block Elements, and Composition Objects. Each structure has Slack's API
//! documentation copied in-place so you don't have to leave your editor to
//! remember the details of the block kit API.
//!
//! Every model has builders that leverage Rust's type system
//! to help you provide every required field, so you can be confident in your app.
//!
//! ## Troubleshooting common compiler errors
//! `Method build not found for ...Builder` - Dig into the error message,
//! you'll find something like `RequiredMethodNotCalled<method::foo>`,
//! meaning you need to call `.foo()` before you can call `.build()`!
//!
//! # Example
//! Using an example from Slack's Documentation:
//! ```json
//! {
//!   "type": "section",
//!   "text": {
//!     "text": "*Sally* has requested you set the deadline for the Nano launch project",
//!     "type": "mrkdwn"
//!   },
//!   "accessory": {
//!     "type": "datepicker",
//!     "action_id": "datepicker123",
//!     "initial_date": "1990-04-28",
//!     "placeholder": {
//!       "type": "plain_text",
//!       "text": "Select a date"
//!     }
//!   }
//! }
//! ```
//!
//! You can use raw Builders like so:
//! ```rust
//! use slack_blocks::{text::ToSlackMarkdown, blocks::Section, elems::DatePicker};
//!
//! let section = Section::builder()
//!                       .text("*Sally* has requested you set the deadline for the Nano launch project".markdown())
//!                       .accessory(DatePicker::builder()
//!                                             .action_id("datepicker123")
//!                                             .initial_date((28, 4, 1990))
//!                                             .placeholder("Select a date")
//!                                             .build()
//!                       )
//!                       .build();
//! ```
//!
//! Or enable the `unstable` feature and use xml macros:
//! ```rust
//! use slack_blocks::blox::*;
//!
//! let pick_date = blox! {
//!   <date_picker action_id="datepicker123"
//!                placeholder="Select a date"
//!                initial_date=(28, 4, 1990) />
//! };
//!
//! let section = blox! {
//!   <section_block accessory=pick_date>
//!     <text kind=plain>"*Sally* has requested you set the deadline for the Nano launch project"</text>
//!   </section_block>
//! };
//! ```
//!
//! Then you can send the block to Slack's API, for example:
//!
//! ```
//! # use slack_blocks::{Block, text::ToSlackMarkdown, blocks::Section, elems::DatePicker};
//! # let section = Section::builder()
//! #                       .text("*Sally* has requested you set the deadline for the Nano launch project".markdown())
//! #                       .accessory(DatePicker::builder()
//! #                                             .action_id("datepicker123")
//! #                                             .initial_date((28, 4, 1990))
//! #                                             .placeholder("Select a date")
//! #                                             .build()
//! #                       )
//! #                       .build();
//! let blocks: Vec<Block> = vec![section.into()]; // using section from examples above
//!
//! reqwest::Client::new()
//!                 .post("https://slack.com/api/chat.postMessage")
//!                 .header("Content-Type", "application/json")
//!                 .bearer_auth("<api token here>")
//!                 .body(serde_json::json!({
//!                   "channel": "<a channel id>",
//!                   "blocks": blocks
//!                 }).to_string())
//!                 .send();
//! ```
//!
//! There is also a crate example (`./examples/reqwest.rs`) that can be run like so:
//! ```sh
//! > export SLACK_API_TOKEN=foo
//! > export SLACK_API_TOKEN=bar
//! > cargo run --all-features --example reqwest
//! ```
//!
//! [Block Kit 🔗]: https://api.slack.com/block-kit
//! [`cargo-make`]: https://github.com/sagiegurari/cargo-make/
//! [issues]: https://github.com/cakekindel/slack-blocks-rs/issues/
//! [Conventional Commits]: https://www.conventionalcommits.org/en/v1.0.0/

#![doc(html_root_url = "https://docs.rs/slack-blocks/0.25.0")]
#![cfg_attr(docsrs, feature(doc_cfg))]
// #![feature(doc_cfg)] // for local docs
#![deny(missing_docs)]
#![cfg_attr(not(test),
            forbid(missing_copy_implementations,
                   missing_debug_implementations,
                   unreachable_pub,
                   unsafe_code,
                   unused_crate_dependencies))]

#[macro_use]
#[cfg(feature = "validation")]
extern crate validator_derive;

#[cfg(feature = "blox")]
#[cfg_attr(docsrs, doc(cfg(feature = "blox")))]
pub mod blox;

pub mod blocks;
pub mod compose;
pub mod elems;

mod build;
#[cfg(feature = "validation")]
mod val_helpr;

#[doc(inline)]
pub use blocks::Block;
#[doc(inline)]
pub use compose::text;
#[doc(inline)]
pub use elems::BlockElement;

mod macros {
  #[macro_export]
  #[doc(hidden)]
  macro_rules! convert {
    (impl From<$source:ty> for $dest:ty => $closure:expr) => {
      impl From<$source> for $dest {
        fn from(src: $source) -> Self {
          $closure(src)
        }
      }
    };
    (impl<'_> From<$source:ident> for $dest:ident => $closure:expr) => {
      impl<'a> From<$source<'a>> for $dest<'a> {
        fn from(src: $source<'a>) -> $dest<'a> {
          $closure(src)
        }
      }
    };
    (impl<$lifetime:lifetime> From<$source:ty> for $dest:ty => $closure:expr) => {
      convert!(impl<$lifetime, > From<$source> for $dest => $closure);
    };
    (impl<$lifetime:lifetime, $($ty_var:ident),*> From<$source:ty> for $dest:ty => $closure:expr) => {
      impl<$lifetime, $($ty_var),*> From<$source> for $dest {
        fn from(src: $source) -> Self {
          $closure(src)
        }
      }
    };
    (impl<$($ty_var:tt),+> From<$source:ty> for $dest:ty => $closure:expr) => {
      impl<$($ty_var),+> From<$source> for $dest {
        fn from(src: $source) -> Self {
          $closure(src)
        }
      }
    };
    (impl From<impl $trait_:ident<$source:ty>> for $dest:ty => $closure:expr) => {
      impl<T> From<T> for $dest where T: $trait_<$source>
      {
        fn from(src: T) -> Self {
          $closure(src)
        }
      }
    };
    (impl<'_> From<impl $trait_:ident<$source:ident>> for $dest:ident => |$param:ident| $body:expr) => {
      impl<'a, T> From<T> for $dest<'a> where T: $trait_<$source<'a>>
      {
        fn from($param: T) -> Self {
          $body
        }
      }
    };
  }
}
